Integrate device agnostic APIs in FSDP library [1/n]

jvandebon opened the pull request:
Summary: For MTIA FSDP support, we need to ensure the FSDP library code handles accelerator devices not limited to CUDA.

Test Plan: CI

Reviewed By: hanzlfs

Differential Revision: D60587415


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

LucasLLC reviewed (approved):
NTE, but stamping because changing cuda specific to generics makes sense to me.

jvandebon committed:
Integrate device agnostic APIs in FSDP library [1/n] (#134337)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134337

For MTIA FSDP support, we need to ensure the FSDP library code handles accelerator devices not limited to CUDA.

Test Plan: CI

Reviewed By: awgu, LucasLLC, hanzlfs

Differential Revision: D60587415
